From 16929ba134adbb7f6da5ff3b9076ff1512454191 Mon Sep 17 00:00:00 2001
From: Maharaja Kennadyrajan <mkenna@qti.qualcomm.com>
Date: Mon, 7 Jul 2025 16:56:30 +0530
Subject: [PATCH] QSDK:CP: wifi: mac80211: set 5GHz bw_limit to 320MHz for EHT
 to enable 240MHz support

EHT connections on the 5GHz band currently do not set a bandwidth limit,
which may prevent the use of 240MHz operation when puncturing is enabled.
The default behavior does not explicitly allow 240MHz, as the bw_limit
remains unset for 5GHz EHT.

Also, the parsing logic for QCN vendor-specific elements incorrectly
casts the pointer when assigning the 240MHz capability structure, and the
conditional block for parsing QCN elements is misplaced, preventing it from
being executed in some cases when EHT 240MHz is enabled.

Set the bw_limit to IEEE80211_CONN_BW_LIMIT_320 for 5GHz EHT connections
to enable 240MHz operation. This ensures that the connection logic
correctly reflects the capabilities of EHT in the 5GHz band. Fix the
pointer cast in the QCN vendor IE parser and move the parsing block outside
the WMM condition to ensure it is always evaluated when applicable.

Signed-off-by: Balamurugan Ramar <bramar@qti.qualcomm.com>
Signed-off-by: Ramasamy Kaliappan <rkaliapp@qti.qualcomm.com>
Signed-off-by: Maharaja Kennadyrajan <mkenna@qti.qualcomm.com>
---
 net/mac80211/mlme.c  |  7 ++++++-
 net/mac80211/parse.c | 14 +++++++-------
 2 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 73874b6..b810c6d 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -5048,6 +5048,7 @@ static bool ieee80211_assoc_config_link(struct ieee80211_link_data *link,
 	struct ieee80211_mgd_assoc_data *assoc_data =
 		sdata->u.mgd.assoc_data ?: sdata->u.mgd.reconf.add_links_data;
 	struct ieee80211_bss_conf *bss_conf = link->conf;
+	struct ieee80211_bss *bss = (void *)cbss->priv;
 	struct ieee80211_local *local = sdata->local;
 	unsigned int link_id = link->link_id;
 	struct ieee80211_elems_parse_params parse_params = {
@@ -5306,7 +5307,8 @@ static bool ieee80211_assoc_config_link(struct ieee80211_link_data *link,
 							    elems->he_cap_len,
 							    elems->eht_cap,
 							    elems->eht_cap_len,
-							    NULL, 0,
+							    &bss->eht_240mhz_capab,
+							    bss->eht_240mhz_len,
 							    link_sta);
 
 			bss_conf->eht_support = link_sta->pub->eht_cap.has_eht;
@@ -5778,6 +5780,9 @@ ieee80211_determine_our_sta_mode(struct ieee80211_sub_if_data *sdata,
 
 	conn->mode = IEEE80211_CONN_MODE_EHT;
 
+	if (is_5ghz)
+		conn->bw_limit = IEEE80211_CONN_BW_LIMIT_320;
+
 	/* check bandwidth */
 	if (is_6ghz &&
 	    eht_cap->eht_cap_elem.phy_cap_info[0] & IEEE80211_EHT_PHY_CAP0_320MHZ_IN_6GHZ)
diff --git a/net/mac80211/parse.c b/net/mac80211/parse.c
index 8da893f..fb9d6e5 100644
--- a/net/mac80211/parse.c
+++ b/net/mac80211/parse.c
@@ -288,7 +288,7 @@ static void ieee802_11_parse_elems_vendor_qcn(const u8 *pos, u8 elen,
 		switch (pos[0]) {
 		case WLAN_OUI_QCN_ATTR_5GHZ_240MHZ_SUPP:
 			if (rem_elen >= 11) {
-				elems->eht_240mhz_capab = (const struct ieee80211_240mhz_vendor_oper *)pos + 2;
+				elems->eht_240mhz_capab = (const struct ieee80211_240mhz_vendor_oper *)(pos + 2);
 				elems->eht_240mhz_len = pos[1];
 			}
 			break;
@@ -446,13 +446,13 @@ _ieee802_11_parse_elems_full(struct ieee80211_elems_parse_params *params,
 						elems->wmm_param = pos;
 						elems->wmm_param_len = elen;
 					}
-				} else if (elen > 6 && pos[0] == 0x8c && pos[1] == 0xfd &&
-					   pos[2] == 0xf0) { /* WLAN_OUI_QCN */
-					if (calc_crc)
-						crc = crc32_be(crc, pos - 2, elen + 2);
-
-					ieee802_11_parse_elems_vendor_qcn(pos, elen, elems);
 				}
+			} else if (elen > 6 && pos[0] == 0x8c && pos[1] == 0xfd &&
+				   pos[2] == 0xf0) { /* WLAN_OUI_QCN */
+				if (calc_crc)
+					crc = crc32_be(crc, pos - 2, elen + 2);
+
+				ieee802_11_parse_elems_vendor_qcn(pos, elen, elems);
 			}
 			break;
 		case WLAN_EID_RSN:
-- 
2.34.1

